mm2src/coins/coin_errors.rs

```diff
@@ -0,0 +1,216 @@
+            expected: format!("{:?}", expected),
+        }
+    }
+
+    pub fn wrong_sender_addr<Found, Expected>(found: Found, expected: Expected) -> ValidatePaymentError
+    where
+        Found: std::fmt::Debug,
+        Expected: std::fmt::Debug,
+    {
+        ValidatePaymentError::WrongSenderAddress {
+            found: format!("{:?}", found),
+            expected: format!("{:?}", expected),
+        }
+    }
+
+    pub fn wrong_token_addr<Found, Expected>(found: Found, expected: Expected) -> ValidatePaymentError
+    where
+        Found: std::fmt::Debug,
+        Expected: std::fmt::Debug,
+    {
+        ValidatePaymentError::WrongTokenAddress {
+            found: format!("{:?}", found),
+            expected: format!("{:?}", expected),
+        }
+    }
+
+    pub fn wrong_swap_id<Found, Expected>(found: Found, expected: Expected) -> ValidatePaymentError
+    where
+        Found: std::fmt::Debug,
+        Expected: std::fmt::Debug,
+    {
+        ValidatePaymentError::WrongSwapId {
+            found: format!("{:?}", found),
+            expected: format!("{:?}", expected),
+        }
+    }
+
+    pub fn wrong_receiver<Found, Expected>(found: Found, expected: Expected) -> ValidatePaymentError
+    where
+        Found: std::fmt::Debug,
+        Expected: std::fmt::Debug,
+    {
+        ValidatePaymentError::WrongReceiver {
+            found: format!("{:?}", found),
+            expected: format!("{:?}", expected),
+        }
+    }
+
+    pub fn wrong_secret_hash<Found, Expected>(found: Found, expected: Expected) -> ValidatePaymentError
+    where
+        Found: std::fmt::Debug,
+        Expected: std::fmt::Debug,
+    {
+        ValidatePaymentError::WrongSecretHash {
+            found: format!("{:?}", found),
+            expected: format!("{:?}", expected),
+        }
+    }
+
+    pub fn wrong_timelock<Found, Expected>(found: Found, expected: Expected) -> ValidatePaymentError
+    where
+        Found: std::fmt::Debug,
+        Expected: std::fmt::Debug,
+    {
+        ValidatePaymentError::WrongTimeLock {
+            found: format!("{:?}", found),
+            expected: format!("{:?}", expected),
+        }
+    }
+
+    pub fn wrong_value<Found, Expected>(found: Found, expected: Expected) -> ValidatePaymentError
+    where
+        Found: std::fmt::Debug,
+        Expected: std::fmt::Debug,
+    {
+        ValidatePaymentError::WrongValue {
+            found: format!("{:?}", found),
+            expected: format!("{:?}", expected),
+        }
+    }
+}
+
+#[derive(Debug, Display)]
+pub enum MyAddressError {
+    UnexpectedDerivationMethod(UnexpectedDerivationMethod),
+    Deprecated(String),
+    InternalError(String),
+}
+
+impl From<UnexpectedDerivationMethod> for MyAddressError {
+    fn from(err: UnexpectedDerivationMethod) -> Self { Self::UnexpectedDerivationMethod(err) }
+}
+
+impl From<String> for MyAddressError {
+    fn from(err: String) -> Self { Self::InternalError(err) }
+}
```

Comment:
Please don't implement `From<String>` for any errror enum. It may lead to incorrect usage of `?` operator.